src/util/readpool/mod.rs

```diff
@@ -0,0 +1,256 @@
+
+pub use self::config::Config;
+pub use self::context::Context;
+pub use self::priority::Priority;
+
+const TICK_INTERVAL_SEC: u64 = 1;
+
+#[derive(Clone)]
+pub struct ReadPool {
+    pool_high: FuturePool<Context>,
+    pool_normal: FuturePool<Context>,
+    pool_low: FuturePool<Context>,
+    max_tasks_high: usize,
+    max_tasks_normal: usize,
+    max_tasks_low: usize,
+}
+
+impl util::AssertSend for ReadPool {}
+impl util::AssertSync for ReadPool {}
+
+impl ReadPool {
+    pub fn new(config: &Config) -> Self {
+        let tick_interval = Duration::from_secs(TICK_INTERVAL_SEC);
+        let build_context_factory = || || Context {};
+        ReadPool {
+            pool_high: FuturePool::new(
+                config.high_concurrency,
+                config.stack_size.0 as usize,
+                "readpool-high",
+                tick_interval,
+                build_context_factory(),
+            ),
+            pool_normal: FuturePool::new(
+                config.normal_concurrency,
+                config.stack_size.0 as usize,
+                "readpool-normal",
+                tick_interval,
+                build_context_factory(),
+            ),
+            pool_low: FuturePool::new(
+                config.low_concurrency,
+                config.stack_size.0 as usize,
+                "readpool-low",
+                tick_interval,
+                build_context_factory(),
+            ),
+            max_tasks_high: config.max_tasks_high,
+            max_tasks_normal: config.max_tasks_normal,
+            max_tasks_low: config.max_tasks_low,
+        }
+    }
+
+    #[inline]
+    fn get_pool_by_priority(&self, priority: Priority) -> &FuturePool<Context> {
+        match priority {
+            Priority::High => &self.pool_high,
+            Priority::Normal => &self.pool_normal,
+            Priority::Low => &self.pool_low,
+        }
+    }
+
+    #[inline]
+    fn get_max_tasks_by_priority(&self, priority: Priority) -> usize {
+        match priority {
+            Priority::High => self.max_tasks_high,
+            Priority::Normal => self.max_tasks_normal,
+            Priority::Low => self.max_tasks_low,
+        }
+    }
+
+    /// Executes a future (generated by the `future_factory`) on specified future pool,
+    /// returning a success future representing the produced value, or a fail future if
+    /// the future pool is full.
+    pub fn future_execute<F, R>(
+        &self,
+        priority: Priority,
+        future_factory: R,
+    ) -> Box<Future<Item = Result<F::Item, F::Error>, Error = Full> + Send>
+    where
+        R: FnOnce(futurepool::ContextDelegators<Context>) -> F + Send + 'static,
+        F: Future + Send + 'static,
+        F::Item: Send + 'static,
+        F::Error: Send + 'static,
+    {
+        let pool = self.get_pool_by_priority(priority);
+        let max_tasks = self.get_max_tasks_by_priority(priority);
+        if pool.get_running_task_count() >= max_tasks {
```

Comment:
We can get `Error::Full` immediately. So How about return `Result<Box<Future<Item = F::Item, Error = F::Error>>>`?

Reply:
Considered previously. In fact your idea can reduce one `box` in my current code, which is good, but I attempted to make a singular return route, so that the caller can be more compact and consistent talking to the code style. For example, in this ([sample 1](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/storage/mod.rs#L546)) the caller can handle the `full` error just like other kind of errors (i.e. in a same pattern), which looks more elegant. [sample 2](https://github.com/breeswish/tikv/blob/wenxuan/readpool-integration/src/server/service/kv.rs#L101) shows that, a single return route can eliminate and unify the error handling code (previously we need [additional code](https://github.com/breeswish/tikv/blob/master/src/server/service/kv.rs#L114) to handle this error). How do think?

Reply:
I think `self.read_pool.future_execute(......).unwrap_or(box future::err(Error::SchedTooBusy))` is better. And `execute_future` can avoid a `box`. If `Box<Future>` is need, do it in business code.

Reply:
Thank you, it make sense. I have updated my code, PTAL again.